### PR TITLE
New folder structure to accommodate Objective-C  

### PR DESCRIPTION
**Pull request by jeff-hykin**
@matter123 I won't merge this in yet because it would break the other merge requests, but I wanted to get your opinion on this new structure.

### CODE REVIEW

File: scripts/generate.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Dir.chdir __dir__ # go to this directory
+Dir.chdir ".." # go up one dir__
+
+for each_dir in Dir["source/languages/**"]
+    Process.wait(Process.spawn("ruby", each_dir+"/generate.rb"))
```
Comment (matter123) [inline scripts/generate.rb]: this should exit with status code 1 if the subprocess exits with status code 1

**Comment by matter123** [comment]
`test/src/` should become `test/source/` to match.

File: source/languages/c/generate.rb [hunk 2]
```diff
@@ -1,6 +1,6 @@
-require_relative '../textmate_tools.rb'
+require_relative '../../textmate_tools.rb'
 require_relative './tokens.rb'
-require_relative '../shared/numeric.rb'
+require_relative '../../shared_patterns/numeric.rb'
```
Comment (matter123) [inline source/languages/c/generate.rb]: with scripts/generate.rb now, would it be better to add /source/ to the load_path and use?:
`require textmate_tools.rb`
`require shared_patterns/numeric.rb`

Comment (jeff-hykin) [reply source/languages/c/generate.rb]: I didn't even know `load_path` existed, thats really cool. I think that'd be a much better solution

**Comment by jeff-hykin** [comment]
> The `tags.txt` files are a generated file. Would it make more sense to move them out of `/source/`?

I was thinking the same thing, I guess we can just have a tags folder and then point them there

**Comment by matter123** [comment]
It looks like `entity.name.type.namespace` came back.

File: source/languages/c/generate.rb (hunk 2 above)
Comment (jeff-hykin) [reply source/languages/c/generate.rb]: Well now that I've implemented it, it seems decently harder to understand whats going on
```
$LOAD_PATH << File.join(File.dirname(__FILE__), "../../") # add source to the load path 
require 'textmate_tools.rb'
require 'repo_specific_helpers.rb'
require 'shared_patterns/numeric.rb'
require 'shared_patterns/predefined_macros.rb'
require_relative 'tokens'
```

Do you think this is any better?
```
source_dir = "../../"
require_relative source_dir + 'textmate_tools.rb'
require_relative source_dir + 'repo_specific_helpers.rb'
require_relative source_dir + 'shared_patterns/numeric.rb'
require_relative source_dir + 'shared_patterns/predefined_macros.rb'
require_relative './tokens.rb'
```

**Comment by jeff-hykin** [comment]
> entity.name.type.names

I think that might be from merging the predefined macros branch, I'll take a look

File: source/languages/c/generate.rb (hunk 2 above)
Comment (matter123) [reply source/languages/c/generate.rb]: Yeah, the second version is better.

**Comment by jeff-hykin** [comment]
Nope it was my mistake on restoring the `using` inside of the attributes

File: scripts/generate.rb [hunk 3]
```diff
@@ -3,7 +3,9 @@
 
 for each_dir in Dir["source/languages/**"]
     exit_code = Process.wait(Process.spawn("ruby", each_dir+"/generate.rb"))
-    if exit_code != 0
```
Comment (matter123) [inline scripts/generate.rb]: Why was this changed?

Comment (jeff-hykin) [reply scripts/generate.rb]: Actually I could use some help on this section. 
There's this really strange bug. When using backticks in ruby or running the command in the command line, I get exit code `0` for the `cpp/generate.rb`.

However, the same command using `system()` or `Process.spawn()` will return a `76XXX` error code like `76101` and every time I run it the error code changes even though the process finishes successfully.

Comment (matter123) [reply scripts/generate.rb]: `Process.wait` returns the process id of the waited on process. use `$?` to get the exit code.

Comment (jeff-hykin) [reply scripts/generate.rb]: Haha thanks, I should've caught that.

File: package.json [hunk 4]
```diff
@@ -18,6 +18,17 @@
...
+        "C++",
+        "cpp",
+        "c",
+        "objective-c",
+        "objective-cpp",
+        "objective-c++",
+        "syntax",
+        "textmate",
+        "highlighting"
+    ],
```
Comment (matter123) [inline package.json]: maybe added color(ing), a couple of other syntax extensions have that (tree sitter, c++ reloaded, yml, syntax highlighter)

File: source/repo_specific_helpers.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+def saveGrammar(grammar)
```
Comment (matter123) [inline source/repo_specific_helpers.rb]: shoulden't this be part of textmate_tools?

Comment (jeff-hykin) [reply source/repo_specific_helpers.rb]: Ideally textmate_tools.rb will become library/ruby gem and mapping it to a specific folder structure would kind of break that.

File: package.json (hunk 4 above)
Comment (jeff-hykin) [reply package.json]: done 👍

File: package.json [hunk 6]
```diff
@@ -1,11 +1,11 @@
 {
     "name": "better-cpp-syntax",
-    "displayName": "Better C++ Syntax",
-    "description": "The bleeding edge of the VS Code C++ syntax",
+    "displayName": "Better Syntax",
+    "description": "The bleeding edge of some VS Code syntaxes",
```
Comment (matter123) [inline package.json]: maybe list the languages or say C-like. "some VS Code syntaxes" is pretty vague.

**Comment by jeff-hykin** [comment]
Thanks for doing that switch to snake case 👍

File: source/languages/c/generate.rb [hunk 7]
```diff
@@ -1,19 +1,23 @@
...
 
 c_grammar = Grammar.new(
+    name: "C",
+    scope_name: "source.c",
+    version: "",
```
Comment (matter123) [inline source/languages/c/generate.rb]: Is no version information okay?

Comment (jeff-hykin) [reply source/languages/c/generate.rb]: I think so. I should actually probably make it optional in the `textmate_tools.rb`. VS Code overwrites the version information whenever they pull it in.

**Comment by matter123** [comment]
Anything else you think should be in this merge?

**Comment by jeff-hykin** [comment]
Yeah, I'm looking at Objective-C++ examples and it still looks pretty broken :/

**Comment by matter123** [comment]
So, how much is sharable between C++ and Objective-C++?

**Comment by jeff-hykin** [comment]
Apparently a lot https://stackoverflow.com/a/4672776
Sadly the Apple link he provided is broken
